com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java

```diff
@@ -1335,24 +1340,29 @@ public void addNode(final ZigBeeNode node) {
             networkNodes.put(node.getIeeeAddress(), node);
         }
 
-        synchronized (this) {
-            if (networkState != ZigBeeTransportState.ONLINE) {
-                return;
-            }
+        if (networkState != ZigBeeTransportState.ONLINE) {
+            return;
+        }
 
-            for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
-                NotificationService.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        listener.nodeAdded(node);
-                    }
-                });
-            }
+        for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
+            NotificationService.execute(new Runnable() {
+                @Override
+                public void run() {
+                    listener.nodeAdded(node);
+                }
+            });
+        }
 
-            if (networkStateSerializer != null) {
-                networkStateSerializer.serialize(this);
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (ZigBeeNetworkManager.this) {
+                    if (networkStateSerializer != null) {
```

Comment:
Wouldn't it be better to put this check outside of the thread?

Reply:
Same as above ...